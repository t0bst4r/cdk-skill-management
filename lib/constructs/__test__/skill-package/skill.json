{
  "manifest": {
    "manifestVersion": "1.0",
    "publishingInformation": {
      "distributionCountries": [],
      "locales": {
        "de-DE": {
          "description": "My Personal Smart Home",
          "examplePhrases": [],
          "keywords": [
            "Smart Home",
            "Lights",
            "Smart Devices"
          ],
          "name": "Test-Smart-Home-Skill",
          "summary": "My Personal Smart Home"
        }
      },
      "testingInstructions": "Sample Testing Instructions."
    },
    "permissions": [
      {
        "name": "alexa::async_event:write"
      }
    ],
    "apis": {
      "smartHome": {
        "protocolVersion": "3",
        "endpoint": {
          "uri": "//override"
        }
      }
    }
  }
}