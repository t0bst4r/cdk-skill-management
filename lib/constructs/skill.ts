import {IResolvable, IResource, Resource} from 'aws-cdk-lib';
import {Construct} from 'constructs';
import {CfnSkill} from 'aws-cdk-lib/alexa-ask';
import {Role, ServicePrincipal} from 'aws-cdk-lib/aws-iam';
import {Asset} from 'aws-cdk-lib/aws-s3-assets';

/**
 * Enumeration for different Alexa Skill types.
 */
export enum SkillType {
  CUSTOM = 'CUSTOM',
  SMART_HOME = 'SMART_HOME',
}

/**
 * Interface representing an Alexa Skill.
 */
export interface ISkill extends IResource {
  /** The unique ID of the Alexa Skill. */
  readonly skillId: string;
  /** The stage of the Alexa Skill. */
  readonly skillStage: string;
  /** The type of the Alexa Skill. */
  readonly skillType: SkillType;
}

/**
 * Base class for the Alexa Skill construct.
 */
abstract class SkillBase extends Resource implements ISkill {
  /**
   * The unique ID of the Alexa Skill.
   */
  public abstract readonly skillId: string;
  /**
   * The stage of the Alexa Skill.
   */
  public abstract readonly skillStage: string;
  /**
   * The type of the Alexa Skill.
   */
  public abstract readonly skillType: SkillType;
}

/**
 * Type alias for Alexa Skill authentication configuration.
 */
export type SkillAuthenticationConfiguration = CfnSkill.AuthenticationConfigurationProperty;
/**
 * Type alias for Alexa Skill package overrides.
 */
export type SkillPackageOverrides = CfnSkill.OverridesProperty;

/**
 * Interface representing an Alexa Skill package.
 */
export interface SkillPackage {
  /** The asset associated with the Alexa Skill package. */
  readonly asset: Asset;
  /** Overrides for the Alexa Skill package. */
  readonly overrides?: SkillPackageOverrides;
}

/**
 * Properties for creating an Alexa Skill.
 */
export interface SkillProps {
  /** The type of the Alexa Skill. */
  readonly skillType: SkillType;
  /** The stage of the Alexa Skill. */
  readonly skillStage: string;
  /** The vendor ID of the Alexa Skill. */
  readonly vendorId: string;
  /** The authentication configuration for the Alexa Skill. */
  readonly authenticationConfiguration: SkillAuthenticationConfiguration | IResolvable;
  /** The package information for the Alexa Skill. */
  readonly skillPackage: SkillPackage;
}

/**
 * Properties for creating an Alexa Skill Instance from Attributes.
 */
export interface SkillAttributes {
  /**
   * The unique ID of the Alexa Skill.
   */
  readonly skillId: string;
  /**
   * The stage of the Alexa Skill.
   */
  readonly skillStage: string;
  /**
   * The type of the Alexa Skill.
   */
  readonly skillType: SkillType;
}

/**
 * Alexa Skill construct for managing an Alexa Skill via CloudFormation.
 */
export class Skill extends SkillBase {
  /**
   * Factory method to create an Alexa Skill from its ID.
   * @param scope - The construct scope.
   * @param id - The construct ID.
   * @param skillId - The ID of the Alexa Skill.
   * @returns The Alexa Skill construct.
   */
  public static fromSkillId(scope: Construct, id: string, skillId: string): ISkill {
    return new (class extends SkillBase {
      readonly skillId: string = skillId;

      get skillType(): SkillType {
        throw new Error('Access to skillType is not supported when using fromSkillId');
      }

      get skillStage(): string {
        throw new Error('Access to skillStage is not supported when using fromSkillId');
      }
    })(scope, id);
  }

  /**
   * Factory method to create an Alexa Skill from its attributes.
   * @param scope - The construct scope.
   * @param id - The construct ID.
   * @param attributes - The attributes of the Alexa Skill.
   * @returns The Alexa Skill construct.
   */
  public static fromAttributes(scope: Construct, id: string, attributes: SkillAttributes): ISkill {
    return new (class extends SkillBase {
      readonly skillId: string = attributes.skillId;
      readonly skillType: SkillType = attributes.skillType;
      readonly skillStage: string = attributes.skillStage;
    })(scope, id);
  }

  private readonly packageRole: Role;
  private readonly resource: CfnSkill;

  /**
   * The unique ID of the Alexa Skill.
   */
  public readonly skillId: string;
  /**
   * The stage of the Alexa Skill.
   */
  public readonly skillStage: string;
  /**
   * The type of the Alexa Skill.
   */
  public readonly skillType: SkillType;

  /**
   * Creates an instance of the Alexa Skill construct.
   * @param scope - The construct scope.
   * @param id - The construct ID.
   * @param props - The Alexa Skill properties.
   */
  constructor(scope: Construct, id: string, props: SkillProps) {
    super(scope, id);

    this.packageRole = new Role(this, 'PackageRole', {
      assumedBy: new ServicePrincipal('alexa-appkit.amazon.com'),
    });

    this.resource = new CfnSkill(this, 'Default', {
      skillPackage: {
        s3Bucket: props.skillPackage.asset.s3BucketName,
        s3Key: props.skillPackage.asset.s3ObjectKey,
        s3BucketRole: this.packageRole.roleArn,
        overrides: {
          ...props.skillPackage.overrides,
          manifest: {
            ...props.skillPackage.overrides?.manifest,
            publishingInformation: {
              category: props.skillType,
              ...props.skillPackage.overrides?.manifest?.publishingInformation,
            },
          },
        },
      },
      vendorId: props.vendorId,
      authenticationConfiguration: props.authenticationConfiguration,
    });
    this.resource.node.addDependency(this.packageRole);
    this.resource.node.addDependency(props.skillPackage.asset);

    this.skillId = this.resource.ref;
    this.skillStage = props.skillStage;
    this.skillType = props.skillType;
  }
}
