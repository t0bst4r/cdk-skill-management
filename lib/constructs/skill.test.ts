import {App, Stack} from 'aws-cdk-lib';
import {Template} from 'aws-cdk-lib/assertions';
import {Skill, SkillType} from './skill';
import {StringParameter} from 'aws-cdk-lib/aws-ssm';
import {Secret} from 'aws-cdk-lib/aws-secretsmanager';
import {Asset} from 'aws-cdk-lib/aws-s3-assets';
import {Function} from 'aws-cdk-lib/aws-lambda';
import * as path from 'path';

test('Skill', () => {
  const app = new App();
  const stack = new Stack(app, 'TestStack');

  const vendorId = StringParameter.fromStringParameterName(stack, 'VendorId', '/Skill/VendorId');
  const skillCredentials = Secret.fromSecretNameV2(stack, 'SkillCredentials', '/Skill/Credentials');
  const skillPackage = new Asset(stack, 'SkillPackage', {
    path: path.join(__dirname, '__test__', 'skill-package'),
  });
  const skillEndpoint = Function.fromFunctionName(stack, 'SkillEndpoint', 'MySkillEndpointFunction');

  new Skill(stack, 'Skill', {
    skillStage: 'development',
    skillType: SkillType.CUSTOM,
    vendorId: vendorId.stringValue,
    authenticationConfiguration: {
      clientId: skillCredentials.secretValueFromJson('clientId').unsafeUnwrap(),
      clientSecret: skillCredentials.secretValueFromJson('clientSecret').unsafeUnwrap(),
      refreshToken: skillCredentials.secretValueFromJson('refreshToken').unsafeUnwrap(),
    },
    skillPackage: {
      asset: skillPackage,
      overrides: {
        manifest: {
          apis: {smartHome: {endpoint: {uri: skillEndpoint.functionArn}}},
        },
      },
    },
  });

  expect(Template.fromStack(stack)).toMatchSnapshot();
});
