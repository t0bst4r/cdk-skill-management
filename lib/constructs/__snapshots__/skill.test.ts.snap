// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skill 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "VendorIdParameter": {
      "Default": "/Skill/VendorId",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Skill": {
      "DependsOn": [
        "SkillPackageRoleD5356B71",
      ],
      "Properties": {
        "AuthenticationConfiguration": {
          "Fn::Join": [
            "",
            [
              "{{resolve:secretsmanager:arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":secretsmanager:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":secret:/Skill/Credentials:SecretString:::}}",
            ],
          ],
        },
        "SkillPackage": {
          "Overrides": {
            "Manifest": {
              "apis": {
                "smartHome": {
                  "endpoint": {
                    "uri": {
                      "Fn::Join": [
                        "",
                        [
                          "arn:",
                          {
                            "Ref": "AWS::Partition",
                          },
                          ":lambda:",
                          {
                            "Ref": "AWS::Region",
                          },
                          ":",
                          {
                            "Ref": "AWS::AccountId",
                          },
                          ":function:MySkillEndpointFunction",
                        ],
                      ],
                    },
                  },
                },
              },
              "publishingInformation": {
                "category": "CUSTOM",
              },
            },
          },
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3BucketRole": {
            "Fn::GetAtt": [
              "SkillPackageRoleD5356B71",
              "Arn",
            ],
          },
          "S3Key": "58abda48478dca03d3892b5160124892bc28484a0c971f7af1dfcaaea280a702.zip",
        },
        "VendorId": {
          "Ref": "VendorIdParameter",
        },
      },
      "Type": "Alexa::ASK::Skill",
    },
    "SkillPackageRoleD5356B71": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "alexa-appkit.amazon.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
